presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt

```diff
@@ -0,0 +1,437 @@
+
+                    override fun onStart(text: String?) {
+                        _state.postValue(TextToSpeechState(SPEAKING, text))
+                    }
+                })
+                if (textToSpeakOnInit != null) {
+                    speakText(textToSpeakOnInit!!)
+                }
+            } else {
+                LOGGER.warn("Language not available for TTS")
+                textToSpeech?.shutdown()
+                textToSpeech = null
+            }
+        } else {
+            LOGGER.warn("Failed to initialize TTS with error code $status")
+            textToSpeech = null
+        }
+    }
+
+    /**
+     * Register the given speech map to be spoken at the times determined by the given countdown and duration.
+     * @param duration the duration of the current step
+     * @param countdown the countdown LiveData for the current step
+     * @param speechMap the map where keys are offsets into the countdown and values are instructions that should be
+     * spoken to the user.
+     */
+    fun registerSpeechesOnCountdown(duration: Duration, countdown: LiveData<Long>, speechMap: Map<String, String>) {
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("register speeches on countdown called,\nduration: $duration\ncountdown: $countdown\n" +
+                    "speechMap: $speechMap")
+        }
+
+        textToSpeech = TextToSpeech(this, this)
+        val canonicalSpeechMap = getCanonicalSpeechMap(speechMap, duration).toMutableMap()
+        val currentCount = countdown.value
+        if (currentCount != null) {
+            // filter out all the speeches that shouldn't be spoken right now, and sort the ones that should in
+            // increasing order of offset.
+            val skippedSpeeches = canonicalSpeechMap.filter { entry ->
+                val currentOffset = duration.seconds - currentCount
+                entry.key < currentOffset && currentOffset - entry.key < speechGracePeriod
+            }.toSortedMap()
+
+            for (speech in skippedSpeeches) {
+                if (!speech.value.second) {
+                    if (LOGGER.isDebugEnabled) {
+                        LOGGER.debug(
+                                "Spoken Instruction: {}, should have been spoken before the service was initialized," +
+                                        " and was still within the grace period.")
+                    }
+                    speakText(speech.value.first)
+                    canonicalSpeechMap[speech.key] = speech.value.copy(second = true)
+                }
+            }
+        }
+
+        futureSpeechData = FutureSpeechData(canonicalSpeechMap, duration, countdown)
+        futureSpeechData?.countdown?.observeForever(::onCountdownChanged)
+    }
+
+    /**
+     * Stops speaking and removes all registered speeches.
+     */
+    fun clear() {
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("clear() called")
+        }
+
+        val tts = textToSpeech
```

Comment:
textToSpeech?.let {
         // Stop speaking and shut down the current text to speech.
        if (it.isSpeaking) {
           it.stop()
        }

        it.shutdown()
        textToSpeech = null
}